pubsub/azuresb: update SendAcks to use Bulk instead of a separate AMQP connection

@vsaroopchand said:

I actually fixed this in the SB SDK; this custom SendAck can use the newly added Bulk Operation. I'll rework SendAck once this is merged in.

https://github.com/Azure/azure-service-bus-go/pull/99

The newly added SendBatchDisposition operation in the SB SDK throws a 410 error on double acks. How should I handle this?

    --- FAIL: TestConformance/TestDoubleAck (29.48s)
        drivertest.go:518: unhandled error link 3213f1e1-c837-4678-8037-91be22e03c14: status code 410 and description: The lock supplied is invalid. Either the lock expired, or the message has already been removed from the queue.

--- a/pubsub/azuresb/azuresb.go
+++ b/pubsub/azuresb/azuresb.go
@@ -59,29 +59,21 @@ import (
 	"net/url"
 	"os"
 	"path"
-	"runtime"
 	"strings"
 	"sync"
 	"time"
 
 	common "github.com/Azure/azure-amqp-common-go/v3"
-	"github.com/Azure/azure-amqp-common-go/v3/cbs"
-	"github.com/Azure/azure-amqp-common-go/v3/rpc"
 	"github.com/Azure/azure-amqp-common-go/v3/uuid"
 	servicebus "github.com/Azure/azure-service-bus-go"
 	"github.com/Azure/go-amqp"
 	"gocloud.dev/gcerrors"
-	"gocloud.dev/internal/useragent"
 	"gocloud.dev/pubsub"
 	"gocloud.dev/pubsub/batcher"
 	"gocloud.dev/pubsub/driver"
 )
 
 const (
-	// https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-amqp-request-response#update-disposition-status
-	dispositionForAck  = "completed"
-	dispositionForNack = "abandoned"
-
 	listenerTimeout = 1 * time.Second
 )
 
@@ -338,9 +330,6 @@ func (*topic) Close() error { return nil }
 type subscription struct {
 	sbSub *servicebus.Subscription
 	opts  *SubscriptionOptions
-
-	linkErr  error     // saved error for initializing amqpLink
-	amqpLink *rpc.Link // nil if linkErr != nil
 }
 
 // SubscriptionOptions will contain configuration for subscriptions.
@@ -375,37 +364,7 @@ func openSubscription(ctx context.Context, sbNs *servicebus.Namespace, sbTop *se
 	if opts == nil {
 		opts = &SubscriptionOptions{}
 	}
-	sub := &subscription{sbSub: sbSub, opts: opts}
-
-	// Initialize a link to the AMQP server, but save any errors to be
-	// returned in ReceiveBatch instead of returning them here, because we
-	// want "subscription not found" to be a Receive time error.
-	host := fmt.Sprintf("amqps://%s.%s/", sbNs.Name, sbNs.Environment.ServiceBusEndpointSuffix)
-	amqpClient, err := amqp.Dial(host,
-		amqp.ConnSASLAnonymous(),
-		amqp.ConnProperty("product", "Go-Cloud Client"),
-		amqp.ConnProperty("version", servicebus.Version),
-		amqp.ConnProperty("platform", runtime.GOOS),
-		amqp.ConnProperty("framework", runtime.Version()),
-		amqp.ConnProperty("user-agent", useragent.AzureUserAgentPrefix("pubsub")),
-	)
-	if err != nil {
-		sub.linkErr = err
-		return sub, nil
-	}
-	entityPath := sbTop.Name + "/Subscriptions/" + sbSub.Name
-	audience := host + entityPath
-	if err = cbs.NegotiateClaim(ctx, audience, amqpClient, sbNs.TokenProvider); err != nil {
-		sub.linkErr = err
-		return sub, nil
-	}
-	link, err := rpc.NewLink(amqpClient, sbSub.ManagementPath())
-	if err != nil {
-		sub.linkErr = err
-		return sub, nil
-	}
-	sub.amqpLink = link
-	return sub, nil
+	return &subscription{sbSub: sbSub, opts: opts}, nil
 }
 
 // IsRetryable implements driver.Subscription.IsRetryable.
@@ -434,24 +393,8 @@ func (s *subscription) ErrorCode(err error) gcerrors.ErrorCode {
 	return code
 }
 
-// partitionAckID is used as the driver.AckID.
-//
-// We use a batch API to ack/nack messages via the LockToken, but AzureSB
-// doesn't support updating messages from different partitions in the same
-// request. We store the PartitionID (which will default to 0 if partitioning
-// isn't enabled) along with the LockToken so that we can group by it during
-// Ack/Nack.
-type partitionAckID struct {
-	PartitionID int16
-	LockToken   *uuid.UUID
-}
-
 // ReceiveBatch implements driver.Subscription.ReceiveBatch.
 func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*driver.Message, error) {
-	if s.linkErr != nil {
-		return nil, s.linkErr
-	}
-
 	// ReceiveOne will block until rctx is Done; we want to return after
 	// a reasonably short delay even if there are no messages. So, create a
 	// sub context for the RPC.
@@ -470,16 +413,10 @@ func (s *subscription) ReceiveBatch(ctx context.Context, maxMessages int) ([]*dr
 				metadata[key] = strVal
 			}
 		}
-		// partitionID is only available if partitioning is enabled; otherwise,
-		// it defaults to 0.
-		var partitionID int16
-		if sbmsg.SystemProperties != nil && sbmsg.SystemProperties.PartitionID != nil {
-			partitionID = *sbmsg.SystemProperties.PartitionID
-		}
 		messages = append(messages, &driver.Message{
 			Body:     sbmsg.Data,
 			Metadata: metadata,
-			AckID:    &partitionAckID{partitionID, sbmsg.LockToken},
+			AckID:    sbmsg.LockToken,
 			AsFunc:   messageAsFunc(sbmsg),
 		})
 		return nil
@@ -508,7 +445,7 @@ func (s *subscription) SendAcks(ctx context.Context, ids []driver.AckID) error {
 		// Ack is a no-op in Receive-and-Delete mode.
 		return nil
 	}
-	return s.updateMessageDispositions(ctx, ids, dispositionForAck)
+	return s.updateMessageDispositions(ctx, ids, servicebus.Complete)
 }
 
 // CanNack implements driver.CanNack.
@@ -524,87 +461,46 @@ func (s *subscription) SendNacks(ctx context.Context, ids []driver.AckID) error 
 	if !s.CanNack() {
 		panic("unreachable")
 	}
-	return s.updateMessageDispositions(ctx, ids, dispositionForNack)
+	return s.updateMessageDispositions(ctx, ids, servicebus.Abort)
 }
 
-// IMPORTANT: This is a workaround to issue message dispositions in bulk which is not supported in the Service Bus SDK.
-func (s *subscription) updateMessageDispositions(ctx context.Context, ids []driver.AckID, disposition string) error {
+func (s *subscription) updateMessageDispositions(ctx context.Context, ids []driver.AckID, status servicebus.MessageStatus) error {
 	if len(ids) == 0 {
 		return nil
 	}
-
-	// Group by partitionID; AzureSB doesn't support updating dispositions for
-	// messages from different partitions.
-	partitions := map[int16][]amqp.UUID{}
+	var lockTokenIDs []*uuid.UUID
 	for _, ackID := range ids {
-		if pid, ok := ackID.(*partitionAckID); ok {
-			lockTokenBytes := [16]byte(*pid.LockToken)
-			partitions[pid.PartitionID] = append(partitions[pid.PartitionID], amqp.UUID(lockTokenBytes))
+		if uid, ok := ackID.(*uuid.UUID); ok {
+			lockTokenIDs = append(lockTokenIDs, uid)
 		}
 	}
-
-	// Update partitions in parallel.
-	const maxConcurrency = 5
-	sem := make(chan struct{}, maxConcurrency)
-	var mu sync.Mutex
-	var errs []string
-	for _, lockTokens := range partitions {
-		sem <- struct{}{}
-		go func(lockTokens []amqp.UUID) {
-			defer func() { <-sem }() // Release the semaphore.
-			if err := s.updateMessageDispositionsInPartition(ctx, lockTokens, disposition); err != nil {
-				mu.Lock()
-				defer mu.Unlock()
-				errs = append(errs, err.Error())
-			}
-		}(lockTokens)
-	}
-	for n := 0; n < maxConcurrency; n++ {
-		sem <- struct{}{}
-	}
-	if len(errs) == 0 {
-		return nil
-	}
-	return fmt.Errorf("Ack/Nack failed: %s", strings.Join(errs, ", "))
-}
-
-// updateMessageDispositionsInPartition assumes lockTokens are all from the
-// same AzureSB partition.
-func (s *subscription) updateMessageDispositionsInPartition(ctx context.Context, lockTokens []amqp.UUID, disposition string) error {
-	value := map[string]interface{}{
-		"disposition-status": disposition,
-		"lock-tokens":        lockTokens,
-	}
-	msg := &amqp.Message{
-		ApplicationProperties: map[string]interface{}{
-			"operation": "com.microsoft:update-disposition",
-		},
-		Value: value,
-	}
-
-	// We're not actually making use of link.Retryable since we're passing 1
-	// here. The portable type will retry as needed.
-	//
-	// We could just use link.RPC, but it returns a result with a status code
-	// in addition to err, and we'd have to check both.
-	_, err := s.amqpLink.RetryableRPC(ctx, 1, 0, msg)
+	err := s.sbSub.SendBatchDisposition(ctx, servicebus.BatchDispositionIterator{
+		LockTokenIDs: lockTokenIDs,
+		Status:       status,
+	})
+	// The error returned from SendBatchDisposition is confusing. It always returns a non-nil
+	// *BatchDispositionError, which holds a list of LockTokenIDs that had errors,
+	// which might be empty (indicating no error).
 	if err == nil {
+		// Unexpected, but clearly no error.
 		return nil
 	}
-	if !isNotFoundErr(err) {
+	bderr, ok := err.(*servicebus.BatchDispositionError)
+	if !ok {
+		// Unexpected, some other kind of error; just return it.
 		return err
 	}
-	// It's a "not found" error, probably due to the message already being
-	// deleted on the server. If we're just acking 1 message, we can just
-	// swallow the error, but otherwise we'll need to retry one by one.
-	if len(lockTokens) == 1 {
+	if bderr == nil || len(bderr.Errors) == 0 {
+		// No actual errors.
 		return nil
 	}
-	for _, lockToken := range lockTokens {
-		value["lock-tokens"] = []amqp.UUID{lockToken}
-		if _, err := s.amqpLink.RetryableRPC(ctx, 1, 0, msg); err != nil && !isNotFoundErr(err) {
-			return err
+	// Ignore "not found" errors, as they are likely re-acks. Return the
+	// first other error (if any).
+	for _, err := range bderr.Errors {
+		if isNotFoundErr(err) {
+			continue
 		}
+		return err
 	}
 	return nil
 }

--- a/pubsub/driver/driver.go
+++ b/pubsub/driver/driver.go
@@ -148,6 +148,8 @@ type Subscription interface {
 	// It is acceptable for SendAcks to be a no-op for drivers that don't
 	// support message acknowledgement.
 	//
+	// Drivers should suppress errors caused by double-acking a message.
+	//
 	// SendAcks may be called concurrently from multiple goroutines.
 	//
 	// Drivers can control the maximum size of ackIDs and the concurrency

